src/lib/arch/aarch64/Instruction_execute.cc

```diff
@@ -182,6 +182,126 @@ bool conditionHolds(uint8_t cond, uint8_t nzcv) {
+  return result;
+}
+
+/** Unpacks 32-bit floating point number into an array of components.
+ * Zeroth element is sign bit, first element is exponent, and second element is
+ * mantissa.
+ */
+std::array<uint64_t, 3> fp32Unpack(float x) {
+  uint32_t fp;
+  memcpy(&fp, &x, sizeof(float));
+  uint64_t sign = fp >> 31;
+  uint64_t exp = (fp >> 23) & 0xff;
+  uint64_t mantissa = (fp & ((1ULL << 23) - 1));
+  return {sign, exp, mantissa};
+}
+
+/** Unpacks 64-bit floating point number into an array of components.
+ * Zeroth element is sign bit, first element is exponent, and second element is
+ * mantissa.
+ */
+std::array<uint64_t, 3> fp64Unpack(double x) {
+  uint64_t fp;
+  memcpy(&fp, &x, sizeof(double));
+  uint64_t sign = fp >> 63;
+  uint64_t exp = (fp >> 52) & 0x7ff;
+  uint64_t mantissa = fp & ((1ULL << 52) - 1);
+  return {sign, exp, mantissa};
+}
+
+/** Returns estimate of reciprocal square root of floating point number 'x'
+ * based on pseudo code from "Arm Architecture Reference Manual Armv8, for
+ * Armv8-A architecture profile". The type of argument 'x' must be either float
+ * or double.
+ */
+template <typename T>
+T FPRSqrtEstimate(T x) {
+  T result;
+  if (std::isnan(x) || x < 0) {
+    result = std::numeric_limits<T>::quiet_NaN();
+  } else if (x == 0) {
+    result = std::numeric_limits<T>::infinity();
+  } else if (std::isinf(x)) {
+    result = 0;
+  } else {
+    result = 0;
+    // unpacking floating point
+    uint64_t fraction;
+    uint64_t exp;
+    std::array<uint64_t, 3> unpacked;
+    if (sizeof(T) == sizeof(float)) {
+      unpacked = fp32Unpack(x);
+      fraction = unpacked[2] << 29;
+    } else {
+      unpacked = fp64Unpack(x);
+      fraction = unpacked[2];
+    }
+    exp = unpacked[1];
+
+    // handling subnormal
+    if (exp == 0) {
+      while ((fraction >> 51) == 0) {
+        fraction = (fraction << 1) & ((1ULL << 52) - 1);
+        exp -= 1;
+      }
+      fraction = (fraction << 1) & ((1ULL << 52) - 1);
+    }
+
+    uint64_t scaled;
+    if ((exp & 1ULL) == 0) {
+      scaled = 0b100000000 | (fraction >> 44);
+    } else {
+      scaled = 0b010000000 | (fraction >> 45);
+    }
+
+    uint64_t result_exp;
+    if (sizeof(T) == sizeof(float)) {
+      result_exp = (380 - exp) >> 1;
+    } else {
+      result_exp = (3068 - exp) >> 1;
+    }
+    uint64_t estimate = recipSqrtEstimate(scaled);
+
+    if (sizeof(T) == sizeof(float)) {
+      uint32_t packed = (result_exp << 23) | ((estimate & 0xff) << 15);
+      assert(sizeof(packed) == sizeof(result) &&
+             "size of packed bits are different with floating point type");
+      memcpy(&result, &packed, sizeof(T));
+    } else {
+      uint64_t packed = (result_exp << 52) | ((estimate & 0xff) << 44);
+      assert(sizeof(packed) == sizeof(result) &&
+             "size of packed bits are different with floating point type");
+      memcpy(&result, &packed, sizeof(T));
+    }
+  }
+  return result;
+}
+
```

Comment:
I was wondering if an adapted version of the Quake-3 Fast inverse-square root function could be used here? By no means is it any more readable (probably far less so) but it is only about 10 lines of code (and very quick). Also unsure of how its accuracy compares to your implementation...

Reply:
The aim of this was to make SimEng to calculate exactly same estimate of reciprocal square root as hardware. So, I think it would be hard to achieve if we use different methods.